File: frontend/lib/suggestions.ts
import { apiFetch } from "@/lib/api-client";

/** Mirrors backend/app/schemas/suggestion.py's RepoOut. */
export type SuggestedRepo = {
  name: string;
  full_name: string;
  url: string;
  description: string;
  stars: number;
  language: string | null;
  pushed_at: string | null;
};

/** Mirrors backend/app/schemas/suggestion.py's RepoSuggestionOut — a proactively
 * computed suggestion on the Dashboard's "Suggested for you" card, distinct from the
 * repo cards embedded in a chat message's tool_result (lib/chat.ts's RepoSuggestion). */
export type GithubSuggestion = {
  id: string;
  technology: string;
  repo: SuggestedRepo;
  computed_at: string;
  dismissed: boolean;
  source: "dashboard" | "chat";
};

/** GET /suggestions/github?scope=recent (default) — active suggestions from within the
 * last day or so, what the Dashboard card shows. */
export async function listGithubSuggestions(): Promise<GithubSuggestion[]> {
  return apiFetch<GithubSuggestion[]>("suggestions/github");
}

/** GET /suggestions/github?scope=all — the complete history, dismissed or not, most
 * recent first; feeds the Conversations page's Suggestions section. */
export async function listAllGithubSuggestions(): Promise<GithubSuggestion[]> {
  return apiFetch<GithubSuggestion[]>("suggestions/github?scope=all");
}

export async function dismissGithubSuggestion(id: string): Promise<GithubSuggestion> {
  return apiFetch<GithubSuggestion>(`suggestions/github/${id}/dismiss`, { method: "POST" });
}

File: backend/app/routers/suggestions.py
from typing import Literal
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import CurrentUser
from app.models.repo_suggestion import RepoSuggestion
from app.models.user import User
from app.schemas.suggestion import RepoOut, RepoSuggestionOut
from app.services import suggestion_service

router = APIRouter(prefix="/suggestions", tags=["suggestions"])


def _to_out(s: RepoSuggestion) -> RepoSuggestionOut:
    return RepoSuggestionOut(
        id=s.id,
        technology=s.technology_display,
        repo=RepoOut(**s.repo),
        computed_at=s.computed_at,
        dismissed=s.dismissed,
        source=s.source.value,
    )


@router.get("/github", response_model=list[RepoSuggestionOut])
async def list_github_suggestions(
    scope: Literal["recent", "all"] = Query("recent"),
    user: User = CurrentUser,
    db: AsyncSession = Depends(get_db),
) -> list[RepoSuggestionOut]:
    """`scope=recent` (default, the Dashboard card): active suggestions from within the
    last day or so. `scope=all` (the Conversations page's Suggestions section): the
    complete history, dismissed or not — see suggestion_service.list_suggestions."""
    suggestions = await suggestion_service.list_suggestions(db, user.id, scope=scope)
    return [_to_out(s) for s in suggestions]


@router.post("/github/{suggestion_id}/dismiss", response_model=RepoSuggestionOut)
async def dismiss_github_suggestion(
    suggestion_id: UUID, user: User = CurrentUser, db: AsyncSession = Depends(get_db)
) -> RepoSuggestionOut:
    suggestion = await suggestion_service.dismiss_suggestion(db, user.id, suggestion_id)
    if suggestion is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Suggestion not found")
    return _to_out(suggestion)
